### PR TITLE
Rewrite FileAutoComplete component to fix accessibility issues

### LINKED ISSUE #11300: machines: Installation Source autocomplete is tedious to use for deep folders
Cockpit version: 183
OS: Fedora Server
Page: Machines

I have my ISOs buried pretty deep in subfolders of subfolders, and when typing the path for getting to a CentOS iso, it suggested to autocomplete, but I had to take my hands off the keyboard, use the mouse in the dropdown, then start typing the next subfolder, then using the mouse again etc.

Steps to reproduce

1. Put an ISO in a folder 3 or 4 levels deep, ideally with long names of the folders
2. Create a new machine and select filesystem as source

### LINKED ISSUE #12176: Entries expanded in FIleAutoComplete component are not keyboard navigatable
Page: Machines

This component is for now used in Virtual Machines, but it's reusable.
The dropdown entries should be able to be navigated key keyboard up-down keys.

### PR DESCRIPTION
**Pull request by KKoukiou**
Fixes #11300
Fixes #12176

 - [x] shares commits with PR #12152

### CODE REVIEW

**Comment by KKoukiou** [comment]
@garrett @andreasn Can you take a first look so that I will avoid spending time on this implementation in case you don't think it's correct?
Then I have to fix the tests, I didn't run them at all yet.

**Comment by garrett** [comment]
OK, this is getting awesome.

1. We need to wait for the other PR to land (or port the PF override fixes here, so whichever lands first gets the arrow icon & active color fixes).
2. Using the placeholder text to show what would autocomplete is a really interesting idea! I like it. (I wonder if it impacts screen readers though?)
3. Perhaps we want to try to mitigate the overlapping of the text (which would make the text darker or a little wonky if it's not perfectly aligned) by doing a stroke on the text, with a background color and use paint-order to set the text above the stroke. 
   - `text-stroke` is kind-of universal works in our supported browsers; all need the `-webkit-` prefix (even Firefox), currently... see https://caniuse.com/#feat=text-stroke 
   - See https://developer.mozilla.org/en-US/docs/Web/CSS/paint-order
   - We'd want to do a CSS supports for it, of course, so the text isn't broken in browsers that don't support it... but recent-ish versions all our supported browsers _do_ support it thankfully!
4. The prefill text and current text do not line up for some reason.
   - It's probably fixable with CSS? I wonder if this is a browser and OS specific thing (or not).
      - I'm guessing we need to set the line-height on the `::placeholder` CSS pseudo-element. This should work in our supported browsers, except for Edge, where it needs an `-ms-` prefix (`::-ms-placeholder`). See https://caniuse.com/#feat=css-placeholder. I wonder if setting `line-height` to `inherit` might work? If not, we probably need to set it twice: Once for the element, another time for the element in `.ct-form`.
   - Alternatively, we could set the input element to inline-flex and align the children. That could possibly also work. However, I think targeting the placeholder is probably the cleaner approach, as we're actually stating what we want to achieve there, instead of relying on flex to do our "heavy lifting".
   - Screenshot: 
      Screenshot from 2019-07-16 09-37-06

I guess I should do the additional CSS work listed above to make this work as expected. :smile:

**Comment by KKoukiou** [comment]
> OK, this is getting awesome.
> 
>     1. We need to wait for the other PR to land (or port the PF override fixes here, so whichever lands first gets the arrow icon & active color fixes).

I cherry-picked your two commits here as well, so whichever lands first gets them in.

> 
>     2. Using the place-holder text to show what would auto-complete is a really interesting idea! I like it. (I wonder if it impacts screen readers though?)

The screen readers seems to ignore the blurry text from auto-completion. Do you think this is a bug?
> 
>     3. Perhaps we want to try to mitigate the overlapping of the text (which would make the text darker or a little wonky if it's not perfectly aligned) by doing a stroke on the text, with a background color and use paint-order to set the text above the stroke.
>        
>        * `text-stroke` is kind-of universal works in our supported browsers; all need the `-webkit-` prefix (even Firefox), currently... see https://caniuse.com/#feat=text-stroke
>        * See https://developer.mozilla.org/en-US/docs/Web/CSS/paint-order
>        * We'd want to do a CSS supports for it, of course, so the text isn't broken in browsers that don't support it... but recent-ish versions all our supported browsers _do_ support it thankfully!
>        * Something like this?
>          ```css
>          @supports (paint-order: stroke fill) and ((-webkit-text-stroke: 1px red) or (text-stroke: 1px red)) {
>             .autocomplete > input {
>                -webkit-text-stroke: 2px solid var(--color-bg);
>                text-stroke: 2px solid var(--color-bg);
>                paint-order: stroke fill;
>             }
>          }
>          ```
> 
>     4. The prefill text and current text do not line up for some reason.

This must be something specific to our code, because their example works as expected, see [here](https://rawgit.com/patternfly/patternfly-react/gh-pages/patternfly-3/?knob-Show%20Help=true&knob-Field%20Level%20Help%20Content=Please%20specify%20Country%20code%20%3Cbr%3E%20%3Ca%20target%3D%27_blank%27%20href%3D%27https%3A%2F%2Fcountrycode.org%2F%27%3EClick%20here%20for%20a%20list%20of%20Country%20codes%3C%2Fa%3E&knob-Close%20Popover=true&knob-Show%20Modal=true&knob-Match%20Height=true&selectedKind=patternfly-react%2FForms%20and%20Controls%2FTypeAhead%20Select&selectedStory=Menu%20Alignment&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)
But I don't understand what might be causing it.

>        
>        * It's probably fixable with CSS? I wonder if this is a browser and OS specific thing (or not).
>          
>          * I'm guessing we need to set the line-height on the `::placeholder` CSS pseudo-element. This should work in our supported browsers, except for Edge, where it needs an `-ms-` prefix (`::-ms-placeholder`). See https://caniuse.com/#feat=css-placeholder. I wonder if setting `line-height` to `inherit` might work? If not, we probably need to set it twice: Once for the element, another time for the element in `.ct-form`.
>        * Alternatively, we could set the input element to inline-flex and align the children. That could possibly also work. However, I think targeting the placeholder is probably the cleaner approach, as we're actually stating what we want to achieve there, instead of relying on flex to do our "heavy lifting".
>        * Screenshot:
>          Screenshot from 2019-07-16 09-37-06
> 
> 
> I guess I should do the additional CSS work listed above to make this work as expected. smile

**Comment by KKoukiou** [comment]
@garrett another though, do you believe it would be better to keep the dropdown all the time open when the input is focused? I just find it a bit annoying that the dropdown closes whenever I select something from the list.

**Comment by garrett** [comment]
> The screen readers seems to ignore the blurry text from auto-completion. Do you think this is a bug?

No, as long as they read the input, it's probably intended. It would probably be too verbose if they read both the placeholder and the text itself. (Oh, perhaps this isn't placeholder. I have to dig into it more. Still, I think it's mostly redundant and too verbose.)

> something specific to our code

Yep, most likely. And I'm probably to blame. It's most likely due to something in the PF3 overrides and/or the ct-form stuff.

> do you believe it would be better to keep the dropdown all the time open when the input is focused

Good question. For the file auto complete, it probably needs to work differently from the OS selector, as it's a multi-level browser.

Also: Is there a way to add a class to directories versus files? We may want to indicate a difference stylistically. (There's probably _not_ a way, right?)

**Comment by KKoukiou** [comment]
> > The screen readers seems to ignore the blurry text from auto-completion. Do you think this is a bug?
> 
> No, as long as they read the input, it's probably intended. It would probably be too verbose if they read both the placeholder and the text itself. (Oh, perhaps this isn't placeholder. I have to dig into it more. Still, I think it's mostly redundant and too verbose.)
> 
> > something specific to our code
> 
> Yep, most likely. And I'm probably to blame. It's most likely due to something in the PF3 overrides and/or the ct-form stuff.
> 
> > do you believe it would be better to keep the dropdown all the time open when the input is focused
> 
> Good question. For the file auto complete, it probably needs to work differently from the OS selector, as it's a multi-level browser.

I adjusted the code to keep it open. Please try it and let me know if you prefer this or the previous behavior.

> 
> Also: Is there a way to add a class to directories versus files? We may want to indicate a difference stylistically. (There's probably _not_ a way, right?)

Sorry not, I tried it :( The thing is that as the options for the dropdown it accepts only string, not html elements.
But We should probably open an issue for that for the PF-4 component at least.

**Comment by garrett** [comment]
I have fixed the alignment of the hint text.

I also included some WIP code as a comment that works in Firefox to prevent overlapping of the hint text and the input text, but I couldn't get it to work in Chrome... this is odd, as it's supposed to support these methods (and claims to support them in `@supports`). Simple testcases also work, so I think there might be some bug here or we're doing something slightly wrong. I left the code in if you'd like to look at it and test it.

We should fix or remove the WIP code before merging to master. I can try to get it working more tomorrow.

Meanwhile, can you test to make sure the alignment is correct for you?

Also, something I noticed: When I hit backspace, I get an error instead of an attempt at completion. It feels wrong.

**Comment by KKoukiou** [comment]
> I have fixed the alignment of the hint text.
> 
> I also included some WIP code as a comment that works in Firefox to prevent overlapping of the hint text and the input text, but I couldn't get it to work in Chrome... this is odd, as it's supposed to support these methods (and claims to support them in `@supports`). Simple testcases also work, so I think there might be some bug here or we're doing something slightly wrong. I left the code in if you'd like to look at it and test it.
> 
> We should fix or remove the WIP code before merging to master. I can try to get it working more tomorrow.
> 
> Meanwhile, can you test to make sure the alignment is correct for you?
> 
> Also, something I noticed: When I hit backspace, I get an error instead of an attempt at completion. It feels wrong.

Allignment on firefox works as expected. :)
Also I fixed the backspace issue.

**Comment by martinpitt** [comment]
Putting needswork back as long as this breaks tests so hard.

**Comment by KKoukiou** [comment]
> Putting needswork back as long as this breaks tests so hard.

Oh I missed this test, https://209.132.184.41:8493/logs/pull-12331-20190719-151616-841caaf3-verify-debian-testing/log.html
But I don't seem to see any other related failure. Fixing this one just now.

**Comment by martinpitt** [comment]
@KKoukiou : I actually meant the avocado one (I didn't look at the others yet).

**Comment by marusak** [comment]
Needs to rebase to master since #12367 changed tests names

**Comment by KKoukiou** [comment]
@martinpitt It's missing the html preview of the logs, if you find it annoying to go over them, let me know and I can rebase.
All faiilures look unrelated to me, but please take a look because with many of the failing tests, I am not familiar with.

File: pkg/lib/cockpit-components-file-autocomplete.jsx [hunk 1]
```diff
@@ -19,110 +19,63 @@
...
+        this.onInputChanged = this.onInputChanged.bind(this);
+
         this.debouncedChange = debounce(300, (value) => {
-            let cb = () => {
-                this.pendingCallback = false;
-                let stateUpdate = this.filterFiles(value);
-                this.setState(stateUpdate, () => this.onChangeCallback(value, { error: stateUpdate.error }));
+            let cb = (dirPath) => {
+                this.updateFiles(dirPath == '' ? '/' : dirPath);
             };
```
Comment (martinpitt) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: nitpick: please let's not use these unnecessary { ; }. I. e. `let cb = dirPath => this.updateFiles(...);

**Comment by garrett** [comment]
Remaining issues:
- Overflow of long strings escapes the text input widget
- Overflow of items in the dropdown trigger a horizontal scrollbar (perhaps these items should be forced to wrap or truncate on the left so we see the end of the string, which matters the most in this case?)
- Popup still displays in incorrect cases:
  - On focus
  - When an item is selected
  - When the text input is clicked (perhaps this is just on focus)
- Popup should only show when:
  - Down arrowhead is clicked
  - Typing occurs
  - Arrow keys are used

**Comment by KKoukiou** [comment]
> Remaining issues:
> 
>     * Overflow of long strings escapes the text input widget
> 
>     * Overflow of items in the dropdown trigger a horizontal scrollbar (perhaps these items should be forced to wrap or truncate on the left so we see the end of the string, which matters the most in this case?)
> 
>     * Popup still displays in incorrect cases:
>       
>       * On focus
>       * When an item is selected
>       * When the text input is clicked (perhaps this is just on focus)
> 
>     * Popup should only show when:
>       
>       * Down arrowhead is clicked
>       * Typing occurs
>       * Arrow keys are used
> 
>     * There's a very small delay due to filtering and modifying the DOM on keypress; instead, this should be debounced and throttled by a very tiny amount so typing won't be delayed

I opened https://github.com/cockpit-project/cockpit/issues/12409

**Comment by martinpitt** [comment]
The [chrome failure](https://209.132.184.41:8493/logs/pull-12331-20190724-092323-0361688d-cockpit-project-cockpit-fedora-30-selenium-chrome/log) is stable and reproduces locally as well. I tried to interactively open the page with chromium and firefox, and it works fine. The generated screenshot-setUp-832659.png.html also clearly has the #system_information_hardware_text element.

This needs some debugging tricks with someone familiar with Selenium -- @jscotka?

**Comment by martinpitt** [comment]
Setting this to blocked, as #12152 shares some commits and that is ready to land.

**Comment by martinpitt** [comment]
#12152 landed, so this now needs a rebase.

**Comment by marusak** [comment]
selenium is green, please push force to trigger all tests. Thanks

**Comment by martinpitt** [comment]
Thanks for this great work! It works fine on the React demo page, but it works and looks really confusing in the User Menu → Authentication dialog when you do "Add Key":

auth

The completion list is on top of the input line, so you can never see what you type and what's the actual value.

I haven't tested VM or storage pool creation on the machines page.

**Comment by andreasn** [comment]
I think that's because of `#credentials-dialog .modal-body { max-height: 500px; }`

**Comment by KKoukiou** [comment]
> I think that's because of `#credentials-dialog .modal-body { max-height: 500px; }`

Yes, but we probably need to keep a max height for the ssh keys list.
I will need to tweak it a bit.

**Comment by garrett** [comment]
> I think that's because of #credentials-dialog .modal-body { max-height: 500px; }

1. It's caused because of having an overflow, not a max-height.
2. The overflow shouldn't be on the dialog contents here, but on the list itself.
3. Why is this a fixed arbitrary height?
   - It should be based on the viewport height minus a bit more than the size of the dialog's UI (in rem units).
   - In other words, something like `calc(100vh - 35rem);` (I don't know the actual height; I just guessed here.) This should be on the list, of course. The downside is that this needs to be adjusted when other parts of the dialog might change. Sadly, it's hardcoded too, but it's at least a hardcoded offset based on base font size instead of an arbitrary pixel size.

After checking out and trying the branch here today, I see that #2 is fixed at least. the 500px hardcode is still there. In most cases, it's probably okayish, but it's still not great. It should at least be changed to rems, even if it doesn't take the vertical viewport height into account.

**Pull request merged**